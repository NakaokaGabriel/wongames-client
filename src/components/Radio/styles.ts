import styled, { css } from 'styled-components'
import { RadioProps } from '.'

export const Wrapper = styled.div`
  display: flex;
  align-items: center;
`

export const Input = styled.input`
  ${({ theme }) => css`
    cursor: pointer;
    display: flex;
    align-items: center;
    justify-content: center;
    appearance: none;
    width: 1.8rem;
    height: 1.8rem;
    border: 0.2rem solid ${theme.colors.primary};
    border-radius: 100%;
    transition: background border ${theme.transition.fast};
    position: relative;
    outline: none;

    &:before {
      content: '';
      width: 0.8rem;
      height: 0.8rem;
      background: ${theme.colors.primary};
      border-radius: 100%;
      position: absolute;
      top: 50%;
      left: 50%;
      transform: translate3d(-50%, -50%, 0);
      opacity: 0;
      transition: ${theme.transition.fast};
    }

    &:focus {
      box-shadow: 0 0 0.5rem ${theme.colors.primary};
    }

    &:checked {
      &:before {
        opacity: 1;
      }
    }
  `}
`

export const Label = styled.label<Pick<RadioProps, 'labelColor'>>`
  ${({ theme, labelColor }) => css`
    color: ${theme.colors[labelColor!]};
    cursor: pointer;
    padding-left: ${theme.spacings.xxsmall};
    line-height: 1.8rem;
  `}
`
