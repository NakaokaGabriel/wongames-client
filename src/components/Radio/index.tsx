import { InputHTMLAttributes } from 'react'
import * as S from './styles'

type RadioValue = string | ReadonlyArray<string> | number

export type RadioProps = {
  onCheck?: (value: RadioValue) => void
  label?: string
  labelFor?: string
  labelColor?: 'white' | 'black'
  value?: RadioValue
} & InputHTMLAttributes<HTMLInputElement>

const Radio = ({
  label,
  labelFor = '',
  labelColor = 'white',
  value,
  ...props
}: RadioProps) => (
  <S.Wrapper>
    <S.Input
      type="radio"
      name={labelFor}
      id={labelFor}
      value={value}
      {...props}
    />
    <S.Label htmlFor={labelFor} labelColor={labelColor}>
      {label}
    </S.Label>
  </S.Wrapper>
)

export default Radio
