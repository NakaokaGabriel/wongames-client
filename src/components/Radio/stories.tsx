import { Story, Meta } from '@storybook/react/types-6-0'

import Radio, { RadioProps } from '.'

export default {
  title: 'Form/Radio',
  component: Radio,
  parameters: {
    layout: 'fullscreen',
    backgrounds: {
      default: 'won-dark'
    }
  },
  argTypes: {
    onCheck: { action: 'checked' }
  }
} as Meta

export const Default: Story<RadioProps> = (props) => (
  <>
    <div style={{ margin: 10 }}>
      <Radio
        {...props}
        label="primeiro"
        labelFor="primeiro"
        id="primeiro"
        name="nome"
        value="primeiro"
        defaultChecked
      />
    </div>
    <div style={{ margin: 10 }}>
      <Radio
        {...props}
        label="segundo"
        labelFor="segundo"
        id="segundo"
        name="nome"
        value="segundo"
      />
    </div>
    <div style={{ margin: 10 }}>
      <Radio
        {...props}
        label="terceiro"
        labelFor="terceiro"
        id="terceiro"
        name="nome"
        value="terceiro"
      />
    </div>
  </>
)
