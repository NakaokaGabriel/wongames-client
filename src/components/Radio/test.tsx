import { screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'

import { renderWithTheme } from 'utils/tests/helpers'
import theme from 'styles/theme'

import Radio from '.'

describe('<Radio />', () => {
  it('should render the radio label', () => {
    const { container } = renderWithTheme(
      <Radio label="radio label" labelFor="radioLabel" />
    )

    expect(screen.getByRole('radio')).toBeInTheDocument()
    expect(screen.getByLabelText(/radio label/i)).toBeInTheDocument()
    expect(screen.getByText(/radio label/i)).toHaveAttribute(
      'for',
      'radioLabel'
    )
    expect(container.firstChild).toMatchSnapshot()
  })

  it('should without radio label', () => {
    renderWithTheme(<Radio />)

    expect(screen.queryByLabelText('Radio')).not.toBeInTheDocument()
  })

  it('should render radio with black label color', () => {
    renderWithTheme(
      <Radio labelFor="radioLabel" label="radio label" labelColor="black" />
    )

    expect(screen.getByText(/radio label/i)).toHaveStyle({
      color: theme.colors.black
    })
  })

  it('should dispatch on check event when label status changes', async () => {
    const onCheck = jest.fn()

    renderWithTheme(
      <Radio label="radio label" value="anyValue" onCheck={onCheck} />
    )

    expect(onCheck).not.toHaveBeenCalled()

    userEvent.click(screen.getByRole('radio'))

    await waitFor(() => {
      expect(onCheck).toHaveBeenCalledTimes(1)
    })

    expect(onCheck).toHaveBeenCalledWith('anyValue')
  })

  it('should be accessible with tab ', async () => {
    renderWithTheme(
      <Radio label="radio label" labelFor="Radio" value="anyValue" />
    )

    const radio = screen.getByLabelText(/radio label/i)

    expect(document.body).toHaveFocus()

    userEvent.tab()

    expect(radio).toHaveFocus()
  })
})
