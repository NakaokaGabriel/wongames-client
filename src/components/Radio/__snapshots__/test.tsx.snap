// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Radio /> should render the radio label 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  width: 1.8rem;
  height: 1.8rem;
  border: 0.2rem solid #F231A5;
  border-radius: 100%;
  -webkit-transition: background border 0.1s ease-in-out;
  transition: background border 0.1s ease-in-out;
  position: relative;
  outline: none;
}

.c1:before {
  content: '';
  width: 0.8rem;
  height: 0.8rem;
  background: #F231A5;
  border-radius: 100%;
  position: absolute;
  top: 50%;
  left: 50%;
  -webkit-transform: translate3d(-50%,-50%,0);
  -ms-transform: translate3d(-50%,-50%,0);
  transform: translate3d(-50%,-50%,0);
  opacity: 0;
  -webkit-transition: 0.1s ease-in-out;
  transition: 0.1s ease-in-out;
}

.c1:focus {
  box-shadow: 0 0 0.5rem #F231A5;
}

.c1:checked:before {
  opacity: 1;
}

.c2 {
  color: #FAFAFA;
  cursor: pointer;
  padding-left: 0.8rem;
  line-height: 1.8rem;
}

<div
  class="c0"
>
  <input
    class="c1"
    id="radioLabel"
    name="radioLabel"
    type="radio"
    value=""
  />
  <label
    class="c2"
    for="radioLabel"
  >
    radio label
  </label>
</div>
`
